gensim/models/doc2vec.py

```diff
@@ -585,18 +592,22 @@ def _vocab_from(self, documents, progress_per=10000):
                 interval_start = default_timer()
                 interval_count = total_words
             document_length = len(document.words)
+
             for tag in document.tags:
                 self.docvecs.note_doctag(tag, document_no, document_length)
+
             for word in document.words:
-                total_words += 1
-                if word in vocab:
-                    vocab[word].count += 1
-                else:
-                    vocab[word] = Vocab(count=1)
+                vocab[word] += 1
+
+            if self.max_vocab_size and len(vocab) > self.max_vocab_size:
+                total_words += utils.prune_vocab(vocab, min_reduce)
+                min_reduce += 1
```

Comment:
There's surely a point where escalation is necessary – but only when a prune of the minimum number isn't removing very many words, making another call imminent. In many situations, repeated calls with the same floor will be effective and still practical. 

Especially think of cases where `true_vocab_size < (2 * max_vocab_size)`, and where eliminating all true count=1 words would be enough. A rapidly-escalating floor likely manages just two prunes – good! – but drops a lot more true count=2 words. (And, by creating all that extra headroom, a bunch of true count=1 words, that arrive after the last prune, also survive.) A slower-escalating floor might need ~4-5 prunes, but is much more precise: drops far fewer count=2 words, and lets fewer count=1 words sneak through in the final unpruned batch. 

The first thought that comes to mind is making `min_reduce` escalate if and only if the prune doesn't remove 'enough' words, where 'enough' is something like 1/10th of `max_vocab_size`. Another relevant option might be to apply @mfcabrera's PR #270 count-min-sketch.